### ISSUE TITLE
zstd @1.4.5_0+universal: ar: libzstd.a: Inappropriate file type or format

### ISSUE THREAD

**Issue by thetrial** [facebook/zstd#2169]
In MacPorts zstd 1.4.5 does not build anyore. 1.4.4 built, so this seems to be a regression.

```
:info:build compiling dynamic library 1.4.5
:info:build creating pkgconfig
:info:build compiling static library
:info:build warning: /Library/Developer/CommandLineTools/usr/bin/ranlib: archive library: libzstd.a will be fat and ar(1) will not be able to operate on it
:info:build compiling static library
:info:build ar: libzstd.a is a fat file (use libtool(1) or lipo(1) and ar(1) on it)
:info:build ar: libzstd.a: Inappropriate file type or format
:info:build make[1]: *** [libzstd.a] Error 1
:info:build make[1]: *** Waiting for unfinished jobs....
:info:build creating versioned links
:info:build make[1]: Leaving directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_archivers_zstd/zstd/work/zstd-1.4.5/lib'
:info:build make: *** [lib] Error 2
:info:build make: *** Waiting for unfinished jobs....
:info:build creating versioned links
:info:build make[1]: Leaving directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_archivers_zstd/zstd/work/zstd-1.4.5/lib'
:info:build make: Leaving directory `/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_archivers_zstd/zstd/work/zstd-1.4.5'
:info:build Command failed:  cd "/opt/local/var/macports/build/_opt_local_var_macports_sources_rsync.macports.org_macports_release_tarballs_ports_archivers_zstd/zstd/work/zstd-1.4.5" && /usr/bin/make -j4 -w allmost 
:info:build Exit code: 2
:error:build Failed to build zstd: command execution failed
:debug:build Error code: CHILDSTATUS 18073 2
:debug:build Backtrace: command execution failed
:debug:build     while executing
:debug:build "system {*}$notty {*}$nice $fullcmdstring"
:debug:build     invoked from within
:debug:build "command_exec build"
:debug:build     (procedure "portbuild::build_main" line 8)
:debug:build     invoked from within
:debug:build "$procedure $targetname"
```
See: https://trac.macports.org/ticket/60542

**Comment by Cyan4973** [comment]
As we develop and test a lot on macosx systems,
it's worth noting that no similar issue has ever been experienced with one of our systems.

So I guess the issue seems related to MacPorts itself, which we don't usually employ, or even some local setting.

**Comment by kencu** [comment]
it fails for me when building with parallel jobs (10.13):
```
CC="clang -arch i386 -arch x86_64" CXX="clang++ -arch 1386 -arch x86_64" make -j 16 allmost  V=1
```

but succeeds with only one job.

**Comment by Cyan4973** [comment]
> it fails for me when building with parallel jobs (10.13):

Possibly related to #2176

**Comment by kencu** [comment]
> If MacPorts compilation issue is related to `-j`, it might be possibly related to #2176 (to be tested)

it is -- I disabled parallel building for 1.4.5, and all is well now. We'll enable it again after you have it sorted --thanks!

**Comment by bimbashrestha** [comment]
What is the status on this issue? @kencu, @thetrial do you mind checking if the new dev branch fixes the -j issue now that https://github.com/facebook/zstd/pull/2176 is merged?

**Comment by kencu** [comment]
Sure, I can do that for you... no Apples around there?

**Comment by bimbashrestha** [comment]
> Sure, I can do that for you... no Apples around there?

Haha, we do have apples here:) Just that I don't use MacPorts and wanted to make sure this issue was actually resolved for the people that do.

Appreciate it!

**Comment by Cyan4973** [comment]
Note : in my own tests of #2176, including on my local Mac laptop,
#2176 significantly reduced the risk of build failure when using `-j` parallel build.

But I don't consider it completely clean yet.
`make -j` is clean.
But `make -j all` is not. This one also compiles test projects and a number of `contrib/` ones. In the traces, one can notice some double-compilation of identical objects. It may still work correctly, likely because these parallel compilations produce the same thing, but that's not really "clean". It shows that the dependency graph is not complete, which is indeed difficult to ensure with multiple nested `Makefile`.

**Comment by Cyan4973** [comment]
I believe I fixed the last issue in #2181 (double generation of `lib`),
it was minor but still generated more work than necessary.
